### PR TITLE
Implement Basic Authentication Middleware

### PR DESCRIPTION
**Pull request by sajosam**
This pull request introduces a basic authentication middleware to manage user authentication in the application, while ensuring that chat routes remain exempt from the authentication process.

**Changes Made:**
1. Loaded username and password from environment variables for authentication.
2. Developed middleware to handle authentication for required routes.
3. Excluded chat routes from authentication checks to maintain accessibility.

### CODE REVIEW

File: app/api/v1/commons.py [hunk 1]
```diff
@@ -1,5 +1,47 @@
 
 import app.schemas.common as resp_schemas
+from app.schemas.common import LoginData
+import os
+from fastapi import APIRouter, Response, HTTPException, status
+from app.utils.authlogin import AuthMiddleware
+
+
```
Comment (agberoz) [inline app/api/v1/commons.py:8]: If you want to us environment variables, define it though config provider

File: app/api/v1/commons.py [hunk 2]
```diff
@@ -1,5 +1,47 @@
...
+AUTH_SERVER = os.getenv("AUTH_SERVER")
+
+
+login = APIRouter()
+
+
```
Comment (agberoz) [inline app/api/v1/commons.py:14]: create a separate python file for auth router

File: app/utils/authlogin.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
+import os
+import jwt
+from jwt import PyJWTError
+from datetime import datetime, timedelta
+from fastapi import Request, HTTPException, status
+import app.schemas.common as resp_schemas
+from fastapi.responses import JSONResponse
+
+
```
Comment (agberoz) [inline app/utils/authlogin.py:9]: initialize class using these parameters

File: app/utils/authlogin.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+SECRET_KEY = os.getenv("SECRET_KEY")
+ALGORITHM = "HS256"
+COOKIE_NAME = "auth_token"
+AUTH_SERVER = os.getenv("AUTH_SERVER")
+
```
Comment (agberoz) [inline app/utils/authlogin.py:14]: Move jwt operation to utils

File: app/utils/authlogin.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+class AuthMiddleware:
+    def __init__(self):
+        if not SECRET_KEY:
+            raise ValueError("SECRET_KEY is missing from environment variables")
+
```
Comment (agberoz) [inline app/utils/authlogin.py:19]: move this method into utils

**Review by agberoz — CHANGES_REQUESTED** [review]

File: app/main.py [hunk 6]
```diff
@@ -112,6 +114,10 @@ def create_app(config):
         allow_headers=["*"],
     )
 
+    # setting instance of auth middleware
+    auth_middleware = AuthMiddleware()
```
Comment (agberoz) [inline app/main.py:118]: add middleware based on a configuration variable

**Review by agberoz — CHANGES_REQUESTED** [review]

File: app/api/v1/auth.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+from fastapi import APIRouter, Response, HTTPException, status
+from app.utils.authlogin import AuthMiddleware
+from app.providers.config import configs
+
+login = APIRouter()
+
+
+@login.post("/")
+def login_user(response: Response, user: LoginData):
+    print(user.username, user.password)
```
Comment (agberoz) [inline app/api/v1/auth.py:12]: no print statements pls

File: app/main.py [hunk 8]
```diff
@@ -112,6 +114,11 @@ def create_app(config):
         allow_headers=["*"],
     )
```
Comment (agberoz) [inline app/main.py:116]: please use meaning full names

configs.auth_enabled and should be bool

File: app/providers/config.py [hunk 9]
```diff
@@ -24,6 +24,12 @@ class Configs(BaseSettings):
     logging_enabled: bool = os.getenv("ENABLE_FILE_LOGGING", False)
     inference_llm_model:str = os.getenv("INFERENCE_LLM_MODEL", "gpt")
 
+    # Auth
```
Comment (agberoz) [inline app/providers/config.py:27]: please use lowercase for variables please

File: app/utils/authlogin.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+from datetime import datetime, timedelta
+from fastapi import Request, status
+import app.schemas.common as resp_schemas
+from fastapi.responses import JSONResponse
+from app.providers.config import configs
+
+
+
+class AuthMiddleware:
+    def __init__(self):
```
Comment (agberoz) [inline app/utils/authlogin.py:13]: don't access configs inside

Comment (agberoz) [reply app/utils/authlogin.py:13]: pass these while initializing

**Review by agberoz — CHANGES_REQUESTED** [review]

File: app/utils/jwt.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+import app.schemas.common as resp_schemas
+from fastapi.responses import JSONResponse
+from app.providers.config import configs
+
+class JWTUtils:
+    def __init__(self, secret_key, algorithm):
+        self.SECRET_KEY = secret_key
+        self.ALGORITHM = algorithm
+
+    def create_access_token(self, data: dict, expires_delta: timedelta = timedelta(minutes=30)):
```
Comment (agberoz) [inline app/utils/jwt.py:14]: create_jwt_token

File: app/utils/jwt.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+        to_encode = data.copy()
+        expire = datetime.now() + expires_delta
+        to_encode.update({"exp": expire})
+        encoded_jwt = jwt.encode(to_encode, self.SECRET_KEY, algorithm=self.ALGORITHM)
+        return encoded_jwt
+
+    def decode_access_token(self, token: str):
```
Comment (agberoz) [inline app/utils/jwt.py:21]: decode jwt token

**Review by agberoz — CHANGES_REQUESTED** [review]

File: app/utils/jwt.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        return encoded_jwt
+
+    def decode_jwt_token(self, token: str):
+        try:
+            payload = jwt.decode(token, self.SECRET_KEY, algorithms=[self.ALGORITHM])
+            return payload
+        except PyJWTError:
+            return None
+
+
```
Comment (agberoz) [inline app/utils/jwt.py:26]: move this into providers/middleware.py

**Review by agberoz — CHANGES_REQUESTED** [review]

File: app/utils/jwt.py (hunk 13 above)
Comment (agberoz) [reply app/utils/jwt.py:26]: verify_token method

File: app/utils/jwt.py [hunk 14]
```diff
@@ -0,0 +1,47 @@
+import jwt
+from jwt import PyJWTError
+from datetime import datetime, timedelta
+from fastapi import Request, status, HTTPException
+from app.providers.config import configs
+
+class JWTUtils:
+    def __init__(self, secret_key, algorithm):
```
Comment (agberoz) [inline app/utils/jwt.py:8]: pass algorithm default value as "HS256"

**Review by agberoz — CHANGES_REQUESTED** [review]

File: app/api/v1/auth.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+        status_code=200,
+        message="Logout Successful",
+        data=None,
+        error=None
+    )
+
+@login.get("/user_info", dependencies=[Depends(verify_token)])
+def get_user_info(request: Request):
+    token = request.cookies.get("auth_token")
+
```
Comment (agberoz) [inline app/api/v1/auth.py:62]: use middleware output here

**Review by agberoz — CHANGES_REQUESTED** [review]

**Comment by agberoz** [comment]
@abhishekas7  loom video ?

**Comment by abhishekas7** [comment]
**Loom video** 

https://www.loom.com/share/c7adf40921b74c07a75fbe29d51cb7b4?sid=d19e084f-42ec-43b4-8f53-d6b9ae64c677

**Comment by abhishekas7** [comment]
**Screenshots**

image
image

**Review by agberoz — DISMISSED** [review]
LGTM

**Comment by abhishekas7** [comment]
Loom Record : https://www.loom.com/share/eed0172af29c4b0697ddcf7f5e9dcb82?sid=bf7e34ce-9f5f-484b-a19f-00a3eadbf15d

**Pull request merged**